subprojects/workers/src/main/java/org/gradle/workers/internal/InProcessCompilerDaemonFactory.java

```diff
@@ -0,0 +1,144 @@
+                BuildOperationWorkerRegistry.Completion workerLease = parentWorkerOperation.operationStart();
+                BuildOperationDetails buildOperation = BuildOperationDetails.displayName(action.getDescription()).parent(parentBuildOperation).build();
+                try {
+                    return buildOperationExecutor.run(buildOperation, new Transformer<DefaultWorkResult, BuildOperationContext>() {
+                        @Override
+                        public DefaultWorkResult transform(BuildOperationContext buildOperationContext) {
+                            return executeInWorkerClassLoader(serverImplementationClass, action, spec, forkOptions.getClasspath(), forkOptions.getSharedPackages());
+                        }
+                    });
+                } finally {
+                    workerLease.operationFinish();
+                }
+            }
+        };
+    }
+
+    private <T extends WorkSpec> DefaultWorkResult executeInWorkerClassLoader(Class<? extends WorkerDaemonProtocol> serverImplementationClass, WorkerDaemonAction<T> action, T spec, Iterable<File> classpath, Iterable<String> sharedPackages) {
+        ClassLoader workerClassLoader = createWorkerClassLoader(serverImplementationClass, classpath, sharedPackages);
+        try {
+            LOGGER.info("Executing {} in in-process worker.", action.getDescription());
+            Callable<?> worker = transferWorkerIntoWorkerClassloader(action, spec, workerClassLoader);
+            Object result = worker.call();
+            DefaultWorkResult workResult = transferResultFromWorkerClassLoader(result);
+            LOGGER.info("Successfully executed {} in in-process worker.", action.getDescription());
+            return workResult;
+        } catch (Exception e) {
+            LOGGER.info("Exception executing {} in in-process worker: {}.", action.getDescription(), e);
+            throw UncheckedException.throwAsUncheckedException(e);
+        }
+    }
+
+    private ClassLoader createWorkerClassLoader(Class<? extends WorkerDaemonProtocol> serverImplementationClass, Iterable<File> classpath, Iterable<String> sharedPackages) {
+        ClassLoader isolatedWorkerClassLoader = classLoaderFactory.createIsolatedClassLoader(new DefaultClassPath(classpath));
+        FilteringClassLoader.Spec isolatedFilteringSpec = new FilteringClassLoader.Spec();
+        for (String sharedPackage : sharedPackages) {
+            isolatedFilteringSpec.allowPackage(sharedPackage);
+        }
+        ClassLoader filteredWorkerClassLoader = classLoaderFactory.createFilteringClassLoader(isolatedWorkerClassLoader, isolatedFilteringSpec);
+        ClassLoader workerServerClassLoader = serverImplementationClass.getClassLoader();
+        return new CachingClassLoader(new MultiParentClassLoader(filteredWorkerClassLoader, workerServerClassLoader));
```

Comment:
I'm a little concerned that we are using this for compiler daemons as well, but that we've removed some of the classloader hierarchy that was there (specifically the groovy and loging classloaders that are a part of that hierarchy).  I'm not sure I have a clear understanding of why that stuff was in there, but we should be sure it's safe to remove.  If not, we should probably have a separate factory for compiler daemons that keeps those classloaders in place.